### ISSUE TITLE
[no-shadow] False positive when declaring any TypeScript enum

### ISSUE THREAD

**Issue by Kenadia** [typescript-eslint/typescript-eslint#2466]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```typescript
export enum SimpleEnum {
  SIMPLE_ENUM_VALUE,
}
```

Full minimal repro here: https://github.com/Kenadia/no-shadow-eslint-enum-test.

**Expected Result**

No error.

**Actual Result**

```
  1:13  error  'SimpleEnum' is already declared in the upper scope  no-shadow
```

**Additional Info**

```
  eslint:cli CLI args: [ 'src/index.ts', '--debug' ] +0ms
  eslint:cli Running on files +3ms
  eslint:config-array-factory Loading JSON config file: /dir/no-shadow-eslint-enum-test/package.json +0ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/dir/no-shadow-eslint-enum-test', loose: false } ] +0ms
  eslint:ignore-pattern   processed: { basePath: '/dir/no-shadow-eslint-enum-test', patterns: [ '/**/node_modules/*' ] } +1ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/dir/no-shadow-eslint-enum-test', loose: false } ] +1ms
  eslint:ignore-pattern   processed: { basePath: '/dir/no-shadow-eslint-enum-test', patterns: [ '/**/node_modules/*' ] } +0ms
  eslint:file-enumerator Start to iterate files: [ 'src/index.ts' ] +0ms
  eslint:file-enumerator File: /dir/no-shadow-eslint-enum-test/src/index.ts +1ms
  eslint:cascading-config-array-factory Load config files for /dir/no-shadow-eslint-enum-test/src. +0ms
  eslint:cascading-config-array-factory No cache found: /dir/no-shadow-eslint-enum-test/src. +0ms
  eslint:config-array-factory Config file not found on /dir/no-shadow-eslint-enum-test/src +3ms
  eslint:cascading-config-array-factory No cache found: /dir/no-shadow-eslint-enum-test. +0ms
  eslint:config-array-factory Loading JS config file: /dir/no-shadow-eslint-enum-test/.eslintrc.js +1ms
  eslint:config-array-factory Config file found: /dir/no-shadow-eslint-enum-test/.eslintrc.js +0ms
  eslint:config-array-factory Loading parser "@typescript-eslint/parser" from /dir/no-shadow-eslint-enum-test/.eslintrc.js +0ms
  eslint:config-array-factory Loaded: @typescript-eslint/parser@4.0.1 (/dir/no-shadow-eslint-enum-test/node_modules/@typescript-eslint/parser/dist/index.js) +1ms
  eslint:config-array-factory Loading plugin "@typescript-eslint" from /dir/no-shadow-eslint-enum-test/.eslintrc.js +215ms
  eslint:config-array-factory Loaded: @typescript-eslint/eslint-plugin@4.0.1 (/dir/no-shadow-eslint-enum-test/node_modules/@typescript-eslint/eslint-plugin/dist/index.js) +1ms
  eslint:config-array-factory Plugin /dir/no-shadow-eslint-enum-test/node_modules/@typescript-eslint/eslint-plugin/dist/index.js loaded in: 122ms +122ms
  eslint:cascading-config-array-factory No cache found: /dir. +340ms
  eslint:config-array-factory Config file not found on /dir +0ms
  eslint:cascading-config-array-factory Stop traversing because of considered root. +0ms
  eslint:rules Loading rule 'no-shadow' (remaining=281) +0ms
  eslint:cascading-config-array-factory Configuration was determined: ConfigArray(2) [ { type: 'config', name: 'DefaultIgnorePattern', filePath: '', criteria: null, env: undefined, globals: undefined, ignorePattern: IgnorePattern { patterns: [Array], basePath: '/dir/no-shadow-eslint-enum-test', loose: false }, noInlineConfig: undefined, parser: undefined, parserOptions: undefined, plugins: undefined, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: undefined, settings: undefined }, { type: 'config', name: '.eslintrc.js', filePath: '/dir/no-shadow-eslint-enum-test/.eslintrc.js', criteria: null, env: { node: true }, globals: undefined, ignorePattern: undefined, noInlineConfig: undefined, parser: { error: null, filePath: '/dir/no-shadow-eslint-enum-test/node_modules/@typescript-eslint/parser/dist/index.js', id: '@typescript-eslint/parser', importerName: '.eslintrc.js', importerPath: '/dir/no-shadow-eslint-enum-test/.eslintrc.js' }, parserOptions: { ecmaVersion: 2018, sourceType: 'module' }, plugins: { '@typescript-eslint': [Object] }, processor: undefined, reportUnusedDisableDirectives: undefined, root: undefined, rules: { 'no-shadow': 'error' }, settings: undefined } ] on /dir/no-shadow-eslint-enum-test/src +3ms
  eslint:ignore-pattern Create with: [ IgnorePattern { patterns: [ '/**/node_modules/*' ], basePath: '/dir/no-shadow-eslint-enum-test', loose: false } ] +346ms
  eslint:ignore-pattern   processed: { basePath: '/dir/no-shadow-eslint-enum-test', patterns: [ '/**/node_modules/*' ] } +0ms
  eslint:ignore-pattern Check {
  filePath: '/dir/no-shadow-eslint-enum-test/src/index.ts',
  dot: false,
  relativePath: 'src/index.ts',
  result: false
} +0ms
  eslint:cli-engine Lint /dir/no-shadow-eslint-enum-test/src/index.ts +0ms
  eslint:linter Linting code for /dir/no-shadow-eslint-enum-test/src/index.ts (pass 1) +0ms
  eslint:linter Verify +0ms
  eslint:linter With ConfigArray: /dir/no-shadow-eslint-enum-test/src/index.ts +0ms
  eslint:linter Generating fixed text for /dir/no-shadow-eslint-enum-test/src/index.ts (pass 1) +22ms
  eslint:source-code-fixer Applying fixes +0ms
  eslint:source-code-fixer shouldFix parameter was false, not attempting fixes +0ms
  eslint:file-enumerator Complete iterating files: ["src/index.ts"] +368ms
  eslint:cli-engine Linting complete in: 369ms +22ms
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.0.1` |
| `TypeScript`                | `4.0.2` |
| `ESLint`                    | `7.8.0` |
| `node`                      | `12.18.3` |

**Comment by bradzacher** [comment]
This is _similar_ to #2447 in that you may not realise it, but enum members create variables.

```ts
const x = 1;
enum Foo {
    x = 2,
    y = x, // what's the value of Foo.y???
}
```

`Foo.y === 2`, because the enum member `x = 2` creates a locally referenceable variable local to the enum.

We can add an option to ignore this.

**Comment by Kenadia** [comment]
Thanks, I did read through #325 thoroughly, so I understand that part of it. In this case I don't see how this could be the intended behavior because this is the entirety of the code:

```typescript
export enum SimpleEnum {
  SIMPLE_ENUM_VALUE,
}
```

**Comment by bradzacher** [comment]
Oh wait I'm so sorry. I completely misunderstood and misread this.
I thought that I fixed this in #2374

Hmm...

**Comment by jsaguet-betclic** [comment]
I had the same issue as @Kenadia.

I realized that I was actually using no-shadow rule from eslint package.
After I switched to the configuration as shown [in the documentation](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-shadow.md), I didn't have any issue anymore:
```
'no-shadow': 'off',
'@typescript-eslint/no-shadow': ['error'],
```

**Comment by jsaguet** [comment]
I had the same issue as @Kenadia.

I realized that I was actually using no-shadow rule from eslint package.
After I switched to the configuration as shown [in the documentation](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-shadow.md), I didn't have any issue anymore:
```
'no-shadow': 'off',
'@typescript-eslint/no-shadow': ['error'],
```

**Comment by Kenadia** [comment]
Got it, this is my mistake then, I'm sorry. I was reading through #325 (led there by Google) and didn't realize that a `@typescript-eslint/no-shadow` rule had been introduced since. Looks like a few people had the same confusion https://github.com/typescript-eslint/typescript-eslint/pull/2374#issuecomment-684728020

**Comment by jsaguet** [comment]
I got confused as well.
It was not very clear to me in the release note that I needed to switch to the new rule `@typescript-eslint/no-shadow` although now that I know it, it does seem clear:

[Release note 4.0.0](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v4.0.0)
> there are some new extension rules that have been added to better support TS that users will need to switch to:
>   - no-redeclare
>   - no-shadow

-> I think it would have been more obvious to mention the full name (@typescript-eslint/no-shadow)

**Comment by bradzacher** [comment]
Thanks for the heads up! I'll update the release note.

**Comment by AArnott** [comment]
When I add:

```json
'@typescript-eslint/no-shadow': ['error'],
```

I get one of these errors for each of my .ts source files:

```
1:1  error  Definition for rule '@typescript-eslint/no-shadow' was not found  @typescript-eslint/no-shadow
```

Why isn't the fix working for me? I believe I'm using the latest versions of all relevant packages:

```json
		"@typescript-eslint/parser": "^4.0.1",
		"eslint": "^7.8.1",
		"eslint-plugin-github": "^4.1.1",
		"eslint-plugin-jest": "^23.20.0",
```

**Comment by bradzacher** [comment]
You need to install `@typescript-eslint/eslint-plugin`

**Comment by AArnott** [comment]
Thanks. Should that be added to the release notes, @bradzacher?

**Comment by bradzacher** [comment]
We generally make the assumption that you're using both the parser and plugin together.
Installing them both is step 0 of our [getting started guide](https://github.com/typescript-eslint/typescript-eslint/tree/master/docs/getting-started/linting).

[We also state in the plugin readme that you should be using the same version of the parser and plugin](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#quick-start).

Your project does have our plugin installed, but it's [installed as a transient dependency via `eslint-plugin-github`](https://github.com/AArnott/pr-autocomplete-app/blob/f739bc2b3f9d636e229b5eaaa851d7e0a7cd866e/yarn.lock#L1914-L1919).

You should always take care to ensure you're only using installed dependencies, and not transient ones.
